### PR TITLE
MONSTER-85 - Read from hidden secondary

### PR DESCRIPTION
**Pull request by alexanderson**

### CODE REVIEW

File: shardmonster/connection.py [hunk 1]
```diff
@@ -82,6 +83,21 @@ def get_connection(cluster_name):
...
+    """The connection client to a hidden secondary in the replica set.
+
+    Connections are cached.
+    Returns None if no hidden secondary exists."""
+    replica_set_conf = connection.admin.command('replSetGetConfig')
+    for member in replica_set_conf['config']['members']:
+        if member['hidden']:
+            uri = 'mongodb://{host}/'.format(host=member['host'])
+            if uri not in _hidden_secondary_connection_cache:
+                _hidden_secondary_connection_cache[uri] = pymongo.MongoClient(uri)
```
Comment (iainb) [inline shardmonster/connection.py]: What problem is the connection cache trying to solve? I don't think this function is called very often, are any timeouts or exceptions that we'd need to deal with if a connection has been idle for a long time (it could be days).

File: shardmonster/sharder.py [hunk 2]
```diff
@@ -228,20 +228,22 @@ def fetch_one(cursor):
...
     if shard_metadata['status'] != metadata.ShardStatus.POST_MIGRATION_DELETE:
         raise Exception('Shard not in delete state')
 
     current_location = shard_metadata['location']
     current_collection = _get_collection_from_location_string(
         current_location, collection_name)
 
-    cursor = current_collection.find({shard_field: shard_key}, {'_id': 1},
-                                     no_cursor_timeout=True)
+    # even though we're deleting the collection we're reading,
```
Comment (iainb) [inline shardmonster/sharder.py]: How can we be sure that all of the documents are deleted if we're reading the ids to delete from a secondary? Its definitely a change, how large of a change I'm not sure.

I'm not sure how big of an issue this is, it feels like an edge case that would only come up if we're working with a small collection that has been frequently modified.

**Comment by iainb** [comment]
`docker-compose run test tox`

```
py2
FAIL: test_delete_after_migration (shardmonster.tests.test_sharder.TestSharder)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/shardmonster/shardmonster/tests/test_sharder.py", line 307, in test_delete_after_migration
    self.assertEqual(0, self.db1.dummy.find({}).count())
AssertionError: 0 != 1
py3
FAIL: test_delete_after_migration (shardmonster.tests.test_sharder.TestSharder)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/shardmonster/shardmonster/tests/test_sharder.py", line 307, in test_delete_after_migration
    self.assertEqual(0, self.db1.dummy.find({}).count())
AssertionError: 0 != 1
```

File: shardmonster/connection.py (hunk 1 above)
Comment (alexanderson) [reply shardmonster/connection.py]: We may not make many of these, but over months...
This was mainly for consistency and so we don't leak connections, but this doesn't (I now realise) have an exipry mechanism. I couldn't use the existing cache as these connections do not have a corresponding cluster name.

File: shardmonster/sharder.py (hunk 2 above)
Comment (alexanderson) [reply shardmonster/sharder.py]: This is linked to your error below. I have now added a commit which does a final delete to catch any last stragglers. This was in the prototype but missed off here.

File: shardmonster/hidden_secondaries.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+def configure_hidden_secondary(cluster_name, host):
+    clusters = connection._get_cluster_coll()
+    return clusters.update_one(
+        {'name': cluster_name},
+        {'$set': {'hidden_secondary_host': host}}
+    )
+
+
+def get_hidden_secondary_connection(cluster_name):
```
Comment (iainb) [inline shardmonster/hidden_secondaries.py]: I'm trying to work out how this should work with only a hidden secondary configured for core01

Reading the code it looks like that any attempts to perform a migration from core02/domain03 would result in a `HiddenSecondarError` being raised from line 45.

I'm probably missing something.

**Comment by iainb** [comment]
Looks like a test flake, only seen it once running the python2 test suite.
```
.........................................................................................F
======================================================================
FAIL: This tests for a bug found during a rollout. The connection for the
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/shardmonster/shardmonster/tests/test_sharder.py", line 374, in test_sync_uses_correct_connection
    self.assertEqual(2, doc2['y'])
AssertionError: 2 != 1

----------------------------------------------------------------------
Ran 116 tests in 26.583s
```

File: shardmonster/sharder.py [hunk 4]
```diff
@@ -219,29 +242,42 @@ def fetch_one(cursor):
...
-            result = current_collection.delete_many({'_id': {'$in': _ids}})
+            result = current_collection.delete_many({
+                '_id': {'$in': _ids},
+                realm['shard_field']: shard_key
+            })
             tum_ti_tum(manager.delete_throttle)
             manager.inc_deleted(by=result.raw_result['n'])
+
+        # if any docs were missed by secondary lag, delete them now
+        result = current_collection.delete_many({
```
Comment (iainb) [inline shardmonster/sharder.py]: I may just be paranoid but this `delete_many` call has an unlimited size. I'd say that it *should* be deleting a small (if not zero) number of docs but there isn't a guarantee. 

I'm thinking that perhaps this `delete_many` should be removed and instead the main loop should run something like:

```
            _delete_source_data(self.collection_name, self.shard_key, self.manager, use_hidden_secondary=False)
            _delete_source_data(self.collection_name, self.shard_key, self.manager, use_hidden_secondary=True)
```
Where both delete phases would use the by-id approach with the same batch size. If no hidden secondary is available it would default to the primary.

What do you think? Am I being overly paranoid? The delete phase currently has the largest impact on performance.

**Comment by iainb** [comment]
found three test failures
py2
```
======================================================================
FAIL: test_delete_after_migration_from_hidden_secondary (shardmonster.tests.test_sharder.TestSharder)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/shardmonster/shardmonster/tests/test_sharder.py", line 318, in test_delete_after_migration_from_hidden_second
ary
    self.assertEqual(0, self.db1.dummy.find({}).count())
AssertionError: 0 != 1

----------------------------------------------------------------------
```
py3
```
ERROR: test_sync_after_copy (shardmonster.tests.test_sharder.TestSharder)
----------------------------------------------------------------------
Traceback (most recent call last):                            
  File "/shardmonster/shardmonster/tests/test_sharder.py", line 280, in test_sync_after_copy
    wait_for_oplog_to_catch_up(db1_hidden_secondary, self.db1)
  File "/shardmonster/shardmonster/tests/base.py", line 15, in wait_for_oplog_to_catch_up
    while oplog_pos(secondary) < oplog_pos(primary):
TypeError: '<' not supported between instances of 'Timestamp' and 'NoneType'
            
====================================================================== 
FAIL: This tests for a bug found during a rollout. The connection for the
----------------------------------------------------------------------                     
Traceback (most recent call last):                                    
  File "/shardmonster/shardmonster/tests/test_sharder.py", line 400, in test_sync_uses_correct_connection
    self.assertEqual(2, doc2['y'])                                    
AssertionError: 2 != 1            
```
